modin/engines/base/frame/data.py

```diff
@@ -546,31 +546,46 @@ def from_labels(self) -> "BasePandasFrame":
             A new BasePandasFrame.
         """
         new_row_labels = pandas.RangeIndex(len(self.index))
+
         # Column labels are different for multilevel index.
-        if len(self.index.names) > 1:
-            # We will also use the `new_column_names` in the calculation of the internal metadata, so this is a
-            # lightweight way of ensuring the metadata matches.
-            new_column_names = pandas.Index(
-                [
-                    self.index.names[i]
-                    if self.index.names[i] is not None
-                    else "level_{}".format(i)
-                    for i in range(len(self.index.names))
-                ]
-            )
-            new_columns = new_column_names.append(self.columns)
+        # We will also use the `new_column_names` in the calculation of the internal metadata, so this is a
+        # lightweight way of ensuring the metadata matches.
+        if self.index.nlevels > 1:
+            level_names = [
+                self.index.names[i]
+                if self.index.names[i] is not None
+                else "level_{}".format(i)
+                for i in range(self.index.nlevels)
+            ]
         else:
-            # See note above about usage of `new_column_names`.
-            new_column_names = pandas.Index(
+            level_names = [
+                self.index.names[0]
+                if self.index.names[0] is not None
+                else "index"
+                if "index" not in self.columns
+                else "level_{}".format(0)
+            ]
+
+        if self.columns.nlevels > 1:
```

Comment:
Is an exception thrown if we do not have the correct labels?

Reply:
It is necessary to ensure unique names for all levels or pandas later throws an exception in `df.reset_index()` in `from_labels_executor`. It is also essential to create `pandas.MultiIndex` object with the same number of levels as `self.column.nlevels`. It could be possible to use fake names, e.g. "level_{#}" and always later replace them with correct ones in query_compiler, but creating a fake `pandas.MultiIndex` object is almost the same code as creating a correct one. It doesn't complicate the code too much and makes `from_labels` operator to output the most common case when `col_level` and `col_fill` are not specified.

Reply:
Also, when creating fake column names, it is always a danger that some of those names already exist in the DataFrame. I figured that I can always forge test column names that would present a name collision, so it is better to use a safe approach repeating Pandas logic.

Reply:
Makes sense, thanks!